### PR TITLE
204 V8 - ListView Variants - Get Published State & UpdateDate for Variant

### PR DESCRIPTION
**Pull request by warrenbuckley**
Connects Umbraco/Umbraco.Private#204

### CODE REVIEW

**Comment by warrenbuckley** [comment]
Ready for you to take a look at please @Shazwazza

File: src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs [hunk 1]
```diff
@@ -88,12 +91,45 @@ internal class ContentMapperProfile : Profile
...
     }
 
+    internal class CultureUpdateDateResolver : IValueResolver<IContent, ContentItemBasic<ContentPropertyBasic>, DateTime>
+    {
+        public DateTime Resolve(IContent source, ContentItemBasic<ContentPropertyBasic> destination, DateTime destMember, ResolutionContext context)
+        {
+            var culture = context.GetCulture();
+
+            //a culture needs to be in the context for a variant content item
+            if (culture == null)
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs]: @warrenbuckley we can't throw here, culture can be null if we are mapping in an invariant context. If it's null it means that it's not been mapped with a culture, so we will have to assume there's no variants being worked on and we'd just return the source.UpdateDate.

File: src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs [hunk 2]
```diff
@@ -88,12 +91,45 @@ internal class ContentMapperProfile : Profile
...
+    }
+
+    internal class CulturePublishedResolver : IValueResolver<IContent, ContentItemBasic<ContentPropertyBasic>, bool>
+    {
+        public bool Resolve(IContent source, ContentItemBasic<ContentPropertyBasic> destination, bool destMember, ResolutionContext context)
+        {
+            var culture = context.GetCulture();
+
+            //a culture needs to be in the context for a variant content item
+            if (culture == null)
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs]: Similar to the other null check we can't throw here

File: src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs [hunk 3]
```diff
@@ -88,12 +91,45 @@ internal class ContentMapperProfile : Profile
...
+    }
+
     internal class CultureNameResolver : IValueResolver<IContent, ContentItemBasic<ContentPropertyBasic>, string>
     {
         public string Resolve(IContent source, ContentItemBasic<ContentPropertyBasic> destination, string destMember, ResolutionContext context)
         {
             var culture = context.GetCulture();
 
+            //a culture needs to be in the context for a variant content item
+            if (culture == null)
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs]: Similar to the other null check we can't throw here

File: src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs [hunk 4]
```diff
@@ -88,12 +91,45 @@ internal class ContentMapperProfile : Profile
...
+    {
+        public DateTime Resolve(IContent source, ContentItemBasic<ContentPropertyBasic> destination, DateTime destMember, ResolutionContext context)
+        {
+            var culture = context.GetCulture();
+
+            //a culture needs to be in the context for a variant content item
+            if (culture == null)
+                return source.UpdateDate;
+
+            var pubDate = source.GetPublishDate(culture);
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs]: When using the variant specific methods like `GetPublishDate` and `GetCultureName`, these should only be used when the content item actually varies by culture else they will return null. So first you need to check `source.ContentType..VariesByCulture()` and if that is true, then you can use GetPublishDate, else you need to use the normal `source.PublishDate` for invariant content.

File: src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs [hunk 5]
```diff
@@ -88,12 +91,45 @@ internal class ContentMapperProfile : Profile
...
     {
         public string Resolve(IContent source, ContentItemBasic<ContentPropertyBasic> destination, string destMember, ResolutionContext context)
         {
             var culture = context.GetCulture();
 
+            //a culture needs to be in the context for a variant content item
+            if (culture == null)
+                return source.Name;
+            
             if (source.CultureNames.TryGetValue(culture, out var name) && !string.IsNullOrWhiteSpace(name))
```
Comment (Shazwazza) [inline src/Umbraco.Web/Models/Mapping/ContentMapperProfile.cs]: see other comment about .VariesByCulture(), same principal here, if it doesn't vary by culture then you need to just return `source.Name`

**Comment by warrenbuckley** [comment]
PR ping pong - back to you @Shazwazza 
PR ping pong

**Pull request merged**